torch/cuda/comm.py

```diff
@@ -120,12 +125,19 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
         inputs, placed on the ``destination`` device.
     """
     output = []
-    itrs = [_take_tensors(tensors, buffer_size) for tensors in inputs]
+    itrs = [_take_tensors(tensors, buffer_size, True) for tensors in inputs]
     for chunks in zip(*itrs):
-        flattened = [_flatten_tensors(chunk) for chunk in chunks]
-        result = reduce_add(flattened, destination)
-        output.extend(_unflatten_tensors(result, chunks[0]))
-    return tuple(output)
+        is_sparse = chunks[0][0].is_sparse
+        if is_sparse:
+            tensors = [chunk[0] for chunk in chunks]
+        else:
```

Comment:
This indexing logic is reaaaaally convoluted. I have no idea what's happening here now. We're checking things 2 levels deep and then we misteriously take the first element from each chunk (I assume it's because they are 1-elem chunks). This piece of code needs more comments